### DISCUSSION TITLE
CS server thread model question

### DISCUSSION THREAD

**Discussion by Yorie** [roflmuffin/CounterStrikeSharp#648] [Q&A] [answered]
Hello, guys,
First of all, thank you for such a huge job on bringing C# to the CS2 server side.

While the code base capabilities are gorgeous, I'm not much familiar with the CS2 codebase and APIs, and documentation of cssharp seems a weak point. So, I'm struggling to find information and hope you could help me with:
- Does CS2 server or cssharp propagates events in sync (or async) mode? I mean, should I care about thread safety if, let's say, two players do some distinct actions at the same time? Or will they come as a single sequence, one by one? I've seen some concurrent collection usages in various plugins, but not sure it is necessary, would like to know more about CS2 server thread and async model.
- The same question is about cssharp global handlers and command listeners?
- What exactly are `Server.NextWorldUpdate` and `Server.NextFrame`? I mean, when and how often the world update happens? When and how often next frame update happens? And how these events overlap/relate with server tick?
- Do you have any plans on improving documentation base of the project? :)

May be somewhere is a good documentation about above questions I've missed?
Would be much appreciate for your help.

**Comment by KillStr3aK** [comment, marked as answer]
Hey there!

> May be somewhere is a good documentation about above questions I've missed?

Our documentation is available at https://docs.cssharp.dev/ also we have example plugins here: [CounterStrikeSharp/tree/main/examples](https://github.com/roflmuffin/CounterStrikeSharp/tree/main/examples)

> I'm not much familiar with the CS2 codebase and APIs

Developers are actively sharing and discuss code snippets in our discord server, this might be a good boost for your knowledge about the game and engine

We indeed miss summaries for a lot of things but they are mostly self explanatory

> Do you have any plans on improving documentation base of the project? :)

Our docs is based on contributions, we always appreciate commits to the docs

> Does CS2 server or cssharp propagates events in sync (or async) mode?
> The same question is about cssharp global handlers and command listeners?

By default, everything runs on the main thread (except some cases, e.g.: plugin (hot)reload, etc..) unless the context is manually switched to an other thread by the developer. (for e.g.: `Task.Run(..)`)

There are no built-in `async` event handlers nor command handlers [yet](https://github.com/roflmuffin/CounterStrikeSharp/issues/218)

> [!WARNING]
> You can only access natives from the **main thread**

What you can do though: once you are in a sync handler, do your async stuff inside `Task.Run(async () => {...})`

_Game event handlers are usually only serves notification purposes_

> What exactly are Server.NextWorldUpdate and Server.NextFrame?

`Server.NextFrame`: Registers a new callback that is scheduled to be executed on the next game frame, if you need tick based scheduling, you can use `Server.RunOnTick`

`Server.NextWorldUpdate`: Registers a new callback that is scheduled to be executed on the next world update, the difference between `GameFrame` and `WorldUpdate` is that the latter is called even during hibernation

the async versions of these methods are just wrapper methods, you can use them in async contexts so you don't have to switch between async and sync contexts

_These are removed from the schedule list after invocation (single use callback)_

**Reply by KillStr3aK** [reply]
you can also ask questions in our discord server
